rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,516 @@
+   remove any such entries.
+4. Create a new checkpoint with:
+   - `id` set to a new v4 uuid
+   - `manifest_id` set to the next manifest version (V + 1)
+   - `checkpoint_expire_time_s` set to None
+   - `metadata` set to `WriterCheckpoint{ epoch: E }`.
+5. Set `writer_epoch` to E
+6. Write the manifest. If this fails with a CAS error, go back to 1.
+
+#### Refreshing the Checkpoint
+
+The writer will periodically re-establish its checkpoint to pin the latest SSTs and SRs. It will run the following
+whenever it detects that the manifest has been updated. This happens either when the manifest is polled
+(at `manifest_poll_interval`), or when the writer detects an update when it goes to update the manifest with it’s local
+view of the core db state:
+1. Read the latest manifest version V. (If the writer is applying an update from the local state then in the first
+   iteration it can assume that it’s local version is the latest. If it’s wrong, the CAS will fail and it should reload
+   the manifest from Object Storage))
+2. Decide whether it needs to re-establish it’s checkpoint. The writer reestablishes the checkpoint whenever the set of
+   L0 SSTs, or Sorted Runs has changed. This includes both changes made locally by the Writer, or by the Compactor.
+    1. If so, create a new checkpoint with
+       - `id` set to a new v4 UUID.
+       - `manifest_id` set to V+1.
+       - `checkpoint_expire_time` set to None.
+       - `metadata` set to `WriterCheckpoint{ epoch: E }`.
+    2. Delete the old checkpoint from `checkpoints`. When we support range scans, we’ll instead want to maintain
+       references to the checkpoint from the range scans and delete the checkpoint when all references are dropped.
+       (We should technically do this for point lookups too, but the likelihood of GC deleting an SST in this case is
+       very low).
+    3. Apply any local changes
+    4. Write the new manifest. If this fails with a CAS error, go back to 1.
+
+#### Clones
+
+The writer will also support initializing a new database from an existing checkpoint. This allows users to “fork” an
+instance of slatedb, allowing it to access all the original db’s data from the checkpoint, but isolate writes to a new
+db instance.
+
+To support this, we add the `Db::open_from_checkpoint` method, which accepts a `parent_path` and `parent_checkpoint`.
+When initializing the db for the first time (detected by the absence of a manifest), the writer will do the following:
+1. Read the current manifest M, which may not be present.
+2. Read the current parent manifest M_p at `parent_path`.
+3. M_p.`initialized` is false, exit with error.
+3. If M is present:
+   1. If the `parent` field is empty, exit with error.
+   2. If `initialized`, then go to step 10.
+4. Compute the clone's checkpoint ID
+   1. If M is present, use M.`parent.parent_checkpoint` as the checkpoint ID
```

Comment:
Nit: Does M being present imply that it is not soft deleted?

Reply:
no, we should check that it's not soft-deleted first. I'll update.